packages/router/src/grpc/offscreen-client.ts

```diff
@@ -15,59 +15,70 @@ const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
 let active = 0;
 
 const activateOffscreen = async () => {
-  const offscreenExists = await chrome.runtime
+  const noOffscreen = chrome.runtime
     .getContexts({
       contextTypes: [chrome.runtime.ContextType.OFFSCREEN_DOCUMENT],
     })
-    .then(contexts => contexts.length > 0);
+    .then(offscreenContexts => !offscreenContexts.length);
 
-  if (!active || !offscreenExists) {
-    await chrome.offscreen.createDocument({
-      url: chrome.runtime.getURL(OFFSCREEN_DOCUMENT_PATH),
-      reasons: [chrome.offscreen.Reason.WORKERS],
-      justification: 'Manages Penumbra transaction WASM workers',
-    });
-  }
-
-  active++;
+  if (!active++ || (await noOffscreen))
+    await chrome.offscreen
+      .createDocument({
+        url: chrome.runtime.getURL(OFFSCREEN_DOCUMENT_PATH),
+        reasons: [chrome.offscreen.Reason.WORKERS],
+        justification: 'Manages Penumbra transaction WASM workers',
+      })
+      .catch((e: unknown) => {
+        // the offscreen window might have been created since we checked
+        // TODO: other failures?
+        console.warn('Failed to create offscreen window', e);
+      });
 };
 
+/**
+ * Decrement and close if there is no remaining activity.
+ */
 const releaseOffscreen = async () => {
-  active--;
-  if (!active) await chrome.offscreen.closeDocument();
+  if (!--active) await chrome.offscreen.closeDocument();
 };
 
 const sendOffscreenMessage = async <T extends OffscreenMessage>(
   req: InternalRequest<T>,
-): Promise<InternalResponse<ActionBuildMessage>> => {
-  try {
-    // Activate the offscreen window by checking if the window already exists,
-    // and if doesn't exist or there aren't any active requests, create it.
-    await activateOffscreen();
-    return await chrome.runtime.sendMessage<InternalRequest<T>, InternalResponse<T>>(req);
-  } finally {
-    // Release the offscreen window resources by closing the offscreen window
-    // after the message is sent, only if there are no active requests.
-    await releaseOffscreen();
-  }
-};
+): Promise<InternalResponse<ActionBuildMessage>> =>
+  chrome.runtime.sendMessage<InternalRequest<T>, InternalResponse<T>>(req);
 
-const buildAction = async (
+/**
+ * Build actions in parallel, in an offscreen window where we can run wasm.
+ * @param cancel Promise that rejects if the build should be cancelled, usually auth denial.
```

Comment:
This pattern is a bit strange to me — can't you just return a promise in this function that `reject`s on failure? Or am I misunderstanding the logic flow here? (Is `cancel` a way to remotely stop `buildActions`, sort of like an `AbortController`'s `signal`? And if so, why not use an `AbortController`? Partly asking just for my own benefit.)

Reply:
Per our convo on Discord, we're not using `AbortController` because you can't use `Promise.race()` with it